base/strings/basic.jl

```diff
@@ -1,57 +1,188 @@
-# TODO: handle other ranges with stride ±1 specially?
-getindex(s::AbstractString, v::AbstractVector{<:Integer}) =
-    sprint(length(v), io->(for i in v; write(io,s[i]) end))
-getindex(s::AbstractString, v::AbstractVector{Bool}) =
-    throw(ArgumentError("logical indexing not supported for strings"))
+See also: `codeunit`, `checkbounds`, `sizeof`, `length`, `endof`
+"""
+ncodeunits(s::AbstractString)
 
-get(s::AbstractString, i::Integer, default) = isvalid(s,i) ? s[i] : default
+"""
+    codeunit(s::AbstractString) -> Type{<:Union{UInt8, UInt16, UInt32}}
+
+Return the code unit type of the given string object. For ASCII, Latin-1, or
+UTF-8 encoded strings, this would be `UInt8`; for UCS-2 and UTF-16 it would be
+`UInt16`; for UTF-32 it would be `UInt32`. The unit code type need not be
+limited to these three types, but it's hard to think of widely used string
+encodings that don't use one of these units. `codeunit(s)` is the same as
+`typeof(codeunit(s,1))` when `s` is a non-empty string.
 
+See also: `ncodeunits`
 """
-    sizeof(s::AbstractString)
+codeunit(s::AbstractString)
+
+"""
+    codeunit(s::AbstractString, i::Integer) -> Union{UInt8, UInt16, UInt32}
+
+Return the code unit value in the string `s` at index `i`. Note that
 
-The number of bytes in string `s`.
+    codeunit(s, i) :: codeunit(s)
+
+I.e. the value returned by `codeunit(s, i)` is of the type returned by
+`codeunit(s)`.
+
+See also: `ncodeunits`, `checkbounds`
+"""
+codeunit(s::AbstractString, i::Integer) = typeof(i) === Int ?
+    throw(MethodError(codeunit, Tuple{typeof(s),Int})) :
+        codeunit(s, Int(i))
+
+"""
+    isvalid(s::AbstractString, i::Integer) -> Bool
+
+Predicate indicating whether the given index is the start of the encoding of
+a character in `s` or not. If `isvalid(s, i)` is true then `s[i]` will return
+the character whose encoding starts at that index, if it's false, then `s[i]`
+will raise an invalid index error. Behavior of `next(s, i)` is similar except
+that the character is returned along with the index of the following character.
+In order for `isvalid(s, i)` to be an O(1) function, the encoding of `s` must
+be [self-synchronizing](https://en.wikipedia.org/wiki/Self-synchronizing_code);
+this is a basic assumption of Julia's generic string support.
+
+See also: `getindex`, `next`, `thisind`, `nextind`, `prevind`, `length`
 
 # Examples
+
 ```jldoctest
-julia> sizeof("❤")
-3
+julia> str = "αβγdef";
+
+julia> isvalid(str, 1)
+true
+
+julia> str[1]
+'α': Unicode U+03b1 (category Ll: Letter, lowercase)
+
+julia> isvalid(str, 2)
+false
+
+julia> str[2]
+ERROR: StringIndexError: invalid character index
+Stacktrace:
+[...]
 ```
 """
-sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary representation")
+isvalid(s::AbstractString, i::Integer) = typeof(i) === Int ?
+    throw(MethodError(isvalid, Tuple{typeof(s),Int})) :
+        isvalid(s, Int(i))
```

Comment:
If there are two methods then any string type that defines `isvalid(s::StringType, i::Integer)` creates an ambiguity. We could not define any `isvalid(s::AbstractString, i::Integer)` method that converts to `Int` but that's annoying because then every type that implements the string interface has to implement the same convert to `Int` logic. This allows subtypes to just implement the `Int` method in the common case or handle `Integer` and either one is fine.